Functions and utility for configuring Group Access Control List based Access

eaaltonen opened the pull request:
This PR adds:
* a number of functions useful for configuring Group based _Access Control List_ (ACL) access on PKCS#11 tokens
* an (_initial_) C++ interface for cktee and ACL functions
* a Command Line utility, `teecli` for easily configuring said accesses

The first commits are minor improvements (IMO), followed by "Add extension function" commit to enable the latter parts, and so on.

## Specific feedback

There are two topics I'm specifically seeking feedback on.

### C++ Interface

Which of the following would best describe the projects stance on C++ code:

* We don't want to include any C++
* An Interface library is fine
* Our users are perfectly capable of managing C++ binary interfaces, it should be a shared library instead.

### Pkg-config Files

The last commit adds `pkg-config` files in the **relocatable** form using `prefix=${pcfiledir}/../..`. This seemed like an ideal option, but afterwards I noticed that pkg-config paths with a `PKG_CONFIG_SYSROOT_DIR` would result in duplicates. We are developing with Yocto where the `PKG_CONFIG_SYSROOT_DIR` is used. I would like to know from developers using the [toolchains](https://optee.readthedocs.io/en/latest/building/toolchains.html) method(s) on which type of file is easiest to use for them:
* the current relocatable files
* the traditional fixed `prefix=/usr` file

## Final notes on the interface

The API calls and pkg-config files have not yet been tested together with software that load `libckteec` dynamically. Some changes to the API and pkg-config files might therefore still be required.

jforissier commented:
Hi @eaaltonen, thanks for the patches. Some comments on a few easy things (for me) first ;-)

- For commit "Use CMake project command to set version":
`Acked-by: Jerome Forissier <jerome@forissier.org>`
A similar change would be nice to have for `libteec/CMakeLists.txt`.
- For commit "Build shared libraries also in CMake":
See a couple of comments inline.
- For commit "Include headers as system headers with full path"
This modifies code that doesn't exist yet so it should be folded into a prior patch in the series.

jforissier reviewed (commented):
Oops forgot to send the inline comments ;-) here they are.

eaaltonen commented on libckteec/CMakeLists.txt:
```diff
@@ -29,7 +29,7 @@ set (SRC
 # Built library
 ################################################################################
-add_library (ckteec ${SRC})
+add_library (ckteec SHARED ${SRC})
```
Most likely buildroot passes the `BUILD_SHARED_LIBS` option.

The arguments in favor of the change are:
* consistency with the Makefile
* doing the right thing by default

However I would consider both of the following to be the right thing:
* build a shared library
* build both a shared and static library (_and this would be even more consistent with the Makefile_).

If you agree to the latter, I can change the commit accordingly.

jforissier replied:
I don't think forcing any value here (shared or static) is correct. See http://lists.buildroot.org/pipermail/buildroot/2020-July/286818.html.

eaaltonen replied:
@jforissier : if I configure it (_with some extra lines_) to build both, would it be ok?

jforissier replied:
Yes, ideally users should be able to build static, shared or both based on some configuration flag. Then every use case and toolchain etc. should be covered. I thought the current code (with no `STATIC` nor `SHARED` keyword here) allowed that, doesn't it?

eaaltonen replied:
At the commit before this PR (d1ae57555d102bc84b1c0761098ced6718a1f785), With CMake version 3.16.3, if I build with
```
mkdir build; cd build
cmake -G Ninja ..
cmake --build .
```
I get only static libraries
```
$ ls *teec/*teec*
libckteec/libckteec.a  libteec/libteec.a
```
and if I delete the build directory and build with
```
mkdir build; cd build
cmake -G Ninja -DBUILD_SHARED_LIBS=TRUE ..
cmake --build .
```
I get only shared libraries and symlinks
```
$ ls *teec/*teec*
libckteec/libckteec.so    libckteec/libckteec.so.0.1.0  libteec/libteec.so.1
libckteec/libckteec.so.0  libteec/libteec.so            libteec/libteec.so.1.0.0
```

and that seems to be CMake's general behaviour.

eaaltonen replied:
@jforissier I rewrote the commit so that it builds both static and shared libraries. Please let me know what you think of it.

Although looking at the change, I'm thinking of moving it to a separate PR. IMO `public/CMakeLists.txt` is declaring things sort of in the wrong direction.

Both @jforissier and @etienne-lms: all current comments addressed or commented on. Waiting for your feedback.

jforissier replied:
> Although looking at the change, I'm thinking of moving it to a separate PR. IMO public/CMakeLists.txt is declaring things sort of in the wrong direction.

Yes, please do.

eaaltonen replied:
Removed CMake library changes from this PR.

eaaltonen commented on libckteec/CMakeLists.txt:
```diff
@@ -62,7 +62,7 @@ target_include_directories(teec
 target_link_libraries (ckteec
 	PRIVATE pthread
-	PRIVATE teec
+	PUBLIC teec
```
I think I'll leave this change away for now. Some of the existing CMakeLists seems partially contradicting.

eaaltonen commented:
> For commit "Include headers as system headers with full path"
This modifies code that doesn't exist yet so it should be folded into a prior patch in the series.

squashed.

> For commit "Use CMake project command to set version":
Acked-by: Jerome Forissier <jerome@forissier.org>
A similar change would be nice to have for libteec/CMakeLists.txt.

Added same change for `libteec` and `Acked-by` line.

etienne-lms commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -0,0 +1,54 @@
+ * the process is a member in `group`.
+ *
+ * @param group	to check membership of.
+ * @return enum rv_groupmember form result.
```
fix indentation, here and few other places.
(here: choose same either tab or space char for indentation)

eaaltonen replied:
@etienne-lms: thanks for the comments. I'll reply as I go and push the changes at the end. Space char seems to be used at the [Doxygen examples](https://www.doxygen.nl/manual/docblocks.html) so I'll go with that.

etienne-lms commented on libckteeaclc/src/group.c:
```diff
@@ -0,0 +1,48 @@
+
+enum rv_groupmember ckteeaclc_current_user_is_member_of(gid_t group)
+{
+	char username[L_cuserid];
```
`= { 0 };`

Init all local variables where defined.
Use `{ 0 }` is universal initializer for structs, arrays, ...

eaaltonen replied:
zero-initialize.

etienne-lms commented on libckteeaclc/src/group.c:
```diff
@@ -0,0 +1,48 @@
+		}
+		ret = getgrouplist(username, group, groups, &grouplistsize);
+		if (ret == -1)
+			return E_GROUPLIST;
```
`free(p_groups);`?

eaaltonen replied:
While looking at this, I noticed the previous `reallocarray` was actually wrong. I fixed this segment to
```
if ((groups = reallocarray(groups, grouplistsize, sizeof(gid_t))) == NULL) {
	free(p_groups);
	return E_MEMORY;
}
ret = getgrouplist(username, group, groups, &grouplistsize);
if (ret == -1) {
	result = E_GROUPLIST;
	goto out;
}
```

etienne-lms reviewed (commented):
Comments for commit "Add libckteeaclc library".

etienne-lms commented on libckteeaclc/CMakeLists.txt:
```diff
@@ -8,16 +8,17 @@ include(GNUInstallDirs)
 # Source files
 ################################################################################
-set (SRC
+set(SRC
```
removing space char is preferred in CMake script files?

eaaltonen replied:
CMake's documentation (for example [target_link_libraries](https://cmake.org/cmake/help/latest/command/target_link_libraries.html#command:target_link_libraries)) does not use them, so I would say yes.

etienne-lms commented on libckteec/src/invoke_ta.c:
```diff
@@ -259,29 +254,53 @@ CK_RV ckteec_invoke_init(void)
+
+	switch (login_method) {
+	case TEEC_LOGIN_PUBLIC:
+		break;
```
could remove this line

eaaltonen replied:
dropped while editing, should be good now.

etienne-lms commented on libckteec/src/invoke_ta.h:
```diff
@@ -6,6 +6,8 @@
 #define LIBCKTEEC_INVOKE_TA_H
 
+#include <inttypes.h>
+#include <grp.h>
```
swap these 2 lines

etienne-lms commented on libckteeaclc/src/group.c:
```diff
@@ -7,6 +7,19 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <errno.h>
```
errno.h before stdio.h.

etienne-lms commented on tee-supplicant/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 project (tee-supplicant C)
 
+add_compile_options(${C_COMPILE_OPTIONS})
```
empty line below?

etienne-lms commented on cli/Makefile:
```diff
@@ -0,0 +1,85 @@
+CXXFLAGS	+= -c -fPIC
+
+DEBUG		?= 0
+ifeq ($(DEBUG), 1)
```
remove space char before `1`.

etienne-lms commented on libckteec/ckteec.pc.in:
```diff
@@ -0,0 +1,14 @@
+includedir=${prefix}/include
+
+Name: ${libname}
+Description: PKCS #11 library in OP-TEE
```
use `PKCS#11`, not `PKCS #11` (no space char). Easier when one greps in the file tree.
few other occurrences in this P-R.

etienne-lms commented on libteec/Makefile:
```diff
@@ -41,7 +48,11 @@ endif
-libteec: $(TEEC_LIBRARY) $(OUT_DIR)/libteec.a
+TEEC_PKGCONF	:= $(OO)/pkgconfig/teec.pc
+
+#$(info $$TEEC_PKGCONF is $(TEEC_PKGCONF))
```
likely to remove :)

eaaltonen replied:
Hmm yes, this is actually something I was planning to remove, but forgot.

etienne-lms reviewed (commented):
various minor comments.

For the integration in **invoke_ta.c**, proposal looks ok, but I think it should be unplugable.
I mean one should be able to use your proposed scheme or a the default one or a new alternate way. Some config switch or like to enable or not at build or runtime.

eaaltonen commented:
My thinking in commit "Add extension function to init with group login" was to modify the previously direct line
```
ckteec_invoke_init
|
init body (`TEEC_InitializeContext` etc.)
```
So that there is a sort of T intersection, where the new entry path `ckteec_invoke_init_with_login` allows specifying the login method in the function parameters, so it looks kind of
```
ckteec_invoke_init
│
├─ckteec_invoke_init_with_login
│
init body (`TEEC_InitializeContext` etc.)
```
The `ckteec_invoke_init_with_login`should easily support at least `TEEC_LOGIN_USER`. I originally had it on the public `ckteec_extensions.h` but reorganized to the current layout after internal PR comments to get the function signatures as clear as possible,

> For the integration in **invoke_ta.c**, proposal looks ok, but I think it should be unplugable.
> I mean one should be able to use your proposed scheme or a the default one or a new alternate way. Some config switch or like to enable or not at build or runtime.

I didn't quite grasp what sort of pluggability you had in mind. Can you clarify a bit?

etienne-lms commented on cli/src/teecli.cpp:
```diff
@@ -0,0 +1,220 @@
+R"(OP-TEE Command Line Interface
+
+Usage:
+  teecli --slot=SLOTNUM init --label=LABEL --set-so-pin=SO-PIN
```
To initialiaze an already SO-pin-inited token, can we use:
`teecli --slot=SLOTNUM init --label=LABEL --so-pin=CUR-PIN --set-so-pin=NEW-PIN`
and for a SO-group-inited token:
`teecli --slot=SLOTNUM init --label=LABEL --so-group=CUR-GROUP --set-so-pin=NEW-PIN`
?

eaaltonen replied:
I started writing to that effect, but `C_InitToken` has only one `pPin` parameter, so that would require adding some non-standard function.

etienne-lms replied:
In that case, can't teecli calls the 2 required API functions C_InitToken (with old SO pin/creds) then C_SetPin (with new SO pin/creds)?
Whatever, one can use teecli twice to do these operation, no issue. Maybe it is even better to have 2 separated calls.

etienne-lms commented on cli/src/teecli.cpp:
```diff
@@ -0,0 +1,220 @@
+	return rv;
+}
+
+CK_RV doInitPin(const cktee::PKCS11Session &p11Session, const cktee::PKCS11Login& /*&p11Login*/, const string &userPin) {
```
why `/*&p11Login*/`?

eaaltonen replied:
I originally had only two parameters, but the login was closed prematurely. This way at least worked as intended.
After that plain `&p11Login` would produce a warning about unused argument.
After that it's a style choice, either:
* not have name at all
* comment the name out

I was more comfortable with the latter. To me it shows more clearly that it's intentionally not used.

etienne-lms replied:
ok, thanks.

etienne-lms commented on cli/src/teecli.cpp:
```diff
@@ -0,0 +1,220 @@
+ * Set user login pin to <tt>userPin</tt> in slot <tt>slotId</tt>.
+ * @param slotId slot number
+ * @param userPin user PIN string
+ * @param soPin SO pin
```
s/pin/PIN/

etienne-lms reviewed (commented):
> > For the integration in invoke_ta.c, proposal looks ok, but I think it should be unplugable.
I mean one should be able to use your proposed scheme or a the default one or a new alternate way. Some config switch or like to enable or not at build or runtime.
> >
> I didn't quite grasp what sort of pluggability you had in mind. Can you clarify a bit?

Looking at the full picture, I don't understand what I asked you.
discard my comment. Currently, enabling or not group/user login relies on optee_os CFG_PKCS11_TA_AUTH_TEE_IDENTITY and that is currently fine.

eaaltonen commented on libckteec/src/invoke_ta.h:
```diff
@@ -83,6 +85,16 @@ static inline CK_RV ckteec_invoke_ctrl_out(unsigned long cmd,
+ * @param login_method such as <tt>TEEC_LOGIN_USER</tt> or <tt>TEEC_LOGIN_GROUP</tt>.
+ * @param gid [optional] used with <tt>TEEC_LOGIN_GROUP</tt> method.
+ */
+CK_RV ckteec_invoke_init_with_login(uint32_t login_method, gid_t gid);
```
Well I did, originally.

However in internal PR review I got several questions about the optional `gid_t gid` argument, which is not used for the `TEEC_LOGIN_PUBLIC` and `TEEC_LOGIN_USER` cases. I resolved the remarks by moving this to the private header and exposing  `ckteec_invoke_init_group_login(gid_t login_gid)` with it's own signature.

If you prefer, I can put it back to the `ckteec_extensions.h` and dump `ckteec_invoke_init_group_login`.

etienne-lms replied:
maybe in extension.h:
```c
CK_RV ckteec_invoke_init_with_login(uint32_t login_method, gid_t login_gid);

static inline CK_RV ckteec_invoke_init_group_login(gid_t login_gid)
{
	return ckteec_invoke_init_with_login(TEEC_LOGIN_GROUP, login_gid);
}

static inline CK_RV ckteec_invoke_init_user_login(void)
{
	return ckteec_invoke_init_with_login(TEEC_LOGIN_USER, 0);
}

static inline CK_RV ckteec_invoke_init_user_public(void)
{
	return ckteec_invoke_init_with_login(TEEC_LOGIN_PUBLIC, 0);
}

eaaltonen replied:
Applied with modifications. I changed the end of the static inline functions to match the order in the constants, that is, `login_<method>`.

eaaltonen commented on libckteec/src/invoke_ta.c:
```diff
@@ -259,29 +254,53 @@ CK_RV ckteec_invoke_init(void)
+		return CKR_ARGUMENTS_BAD;
+	}
+
+	int e = pthread_mutex_lock(&ta_ctx.init_mutex);
```
added.

> And the cpp sources even have other style. That's OK; I think.

Regarding cpp, I interpreted constructors and destructors equivalent to functions (_for braces_). So in that sense I'm trying to apply the OP-TEE coding standards.

eaaltonen commented on libckteec/include/ckteec_extensions.h:
```diff
@@ -0,0 +1,31 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) Vaisala Oyj.
```
In our copyright header it ends with a period. The following `All rights reserved.` has been removed in this PR :)

eaaltonen commented on libckteec/src/invoke_ta.c:
```diff
@@ -259,29 +254,53 @@ CK_RV ckteec_invoke_init(void)
+	case TEEC_LOGIN_USER:
+		break;
+	case TEEC_LOGIN_GROUP:
+		login_data = &login_gid;
```
Doesn't seem strange to me, provided the usage lifetime is within the call.

etienne-lms replied:
ok, fine.

eaaltonen commented on libckteeaclc/Makefile:
```diff
@@ -0,0 +1,86 @@
+
+# Write version to file so that it can be used for up-to-date checks
+MAJ_MIN_P_FILE	:= $(OUT_DIR)/metadata/version
+$(shell mkdir -p $(OUT_DIR)/metadata; echo $(MAJ_MIN_P) | \
```
indented as suggested.

eaaltonen commented on ckteecpp/Makefile:
```diff
@@ -0,0 +1,29 @@
+
+# Write version to file so that it can be used for up-to-date checks
+MAJ_MIN_P_FILE	:= $(OUT_DIR)/metadata/version
+$(shell mkdir -p $(OUT_DIR)/metadata; echo $(MAJ_MIN_P) | \
```
indented.

eaaltonen commented:
@etienne-lms : thank you for the comments. All addressed in some manner.

eaaltonen commented on libckteeaclc/src/cktee_uuid.c:
```diff
@@ -0,0 +1,34 @@
+	if (rv < 0)
+		return rv;
+
+	uuid_t g_uuid = {};
```
This one's missing, just a moment.

eaaltonen replied:
done

etienne-lms commented on ckteecpp/include/ckteecpp/DefaultPKCS11Login.hpp:
```diff
@@ -0,0 +1,88 @@
+	 * @param sessionHandle existing session
+	 * @param userType usually <tt>CK_USER</tt> or <tt>CKU_SO</tt>
+	 * @param loginPin pin code in UTF-8 encoding
+	 * @param logoutRv where to store <tt>C_Logout</tt> return value.
```
It should be about C_Login() return value, no?
s/logoutRv/loginRv/
s/C_Logout/C_Login/

Ditto for other occurrences.

eaaltonen replied:
Same as [this comment](https://github.com/OP-TEE/optee_client/pull/259#discussion_r596191640).

etienne-lms replied:
ok

etienne-lms commented on cli/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+################################################################################
+set(GENERATED_SOURCES_DIR "${CMAKE_BINARY_DIR}/generated-src/teecli")
+file(MAKE_DIRECTORY "${GENERATED_SOURCES_DIR}")
+set(TEECLI_VERSION 0.1.0)
```
s/0.1.0/${PROJECT_VERSION}/ ?

eaaltonen replied:
changed

etienne-lms commented on cli/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+add_executable(teecli ${SRC})
+target_link_libraries(teecli teecli-constants docopt ckteecpp ckteeaclc)
+
+# Install targets
```
missing 
`################################################################################`
above
:-)

eaaltonen replied:
I'm not a fan of banner headers, but if you insist...

etienne-lms replied:
Me neither. This is more about consistency with other inline comments in this file.

jforissier replied:
Same with me, I would certainly approve a later PR to remove the banners, but I agree with @etienne-lms that consistency is good at any time.

etienne-lms commented on cli/src/teecli.cpp:
```diff
@@ -0,0 +1,227 @@
+#include "TeeCliConstants.hpp"
+
+#include <ckteecpp/cktee.hpp>
+#include <ckteeaclc.h>
```
swap the 2 lines above

etienne-lms commented on cli/src/teecli.cpp:
```diff
@@ -0,0 +1,227 @@
+using std::cerr;
+using std::endl;
+using std::map;
+using std::string;
```
i suggest to sort these variable alphabetically. (not a strong opinion)

eaaltonen replied:
sorted.

etienne-lms commented on cli/src/teecli.cpp:
```diff
@@ -0,0 +1,227 @@
+  init                        Initialize token and set Security Officer (SO) login
+  user-login                  Set User login (PIN or group, must match SO login type)
+)";
+/* If SO and user login become independent
```
```
/*
 * If SO and user login become independent

etienne-lms commented on cli/Makefile:
```diff
@@ -22,18 +22,23 @@ endif
+TEECLI_VERSION_FILE	:= $(OUT_DIR)/metadata/version
+$(shell mkdir -p $(OUT_DIR)/metadata; echo $(TEECLI_VERSION) | \
+	diff $(TEECLI_VERSION_FILE) - > /dev/null || \
+	echo $(TEECLI_VERSION) > $(TEECLI_VERSION_FILE) )
```
@jforissier, may I get you attention on this?
It looks ok to me for the intention but I would happily get your experience in makefiles and scripts.

jforissier replied:
Looks OK to me too with the following improvement: replace `$(OUT_DIR)/metadata` with `$(dir $(TEECLI_VERSION_FILE))`.

eaaltonen replied:
Included in next push.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "Use CMake project command to set version"
* "Add function to resolve name to gid"

See my comments, mostly minor stuff. I did not look at the last commit yet.

jforissier commented on libckteec/include/ckteec_extensions.h:
```diff
@@ -0,0 +1,71 @@
+/**
+* ckteec_invoke_init_with_login - Invoke init with an alternative login method.
+*
+* TEE <b>MUST</b> be initiated at this point.
```
I agree. We're not rendering the comment to HTML etc. (Doxygen) so it is more of a distraction IMO.

eaaltonen replied:
I've used the subset of Doxygen that CLion's documentation viewer is able to render correctly.
![CLion_documentation_view](https://user-images.githubusercontent.com/1905693/111500672-ea1d8700-874c-11eb-9279-06126026562b.png)
(_also added a_ `<p>` _there before TEE MUST_).

Please let me know if you still want to discard markup.

eaaltonen replied:
@etienne-lms || @jforissier : please comment.

jforissier replied:
@eaaltonen if that helps you produce some better documentation from source then I'm fine with keeping the markup.

etienne-lms replied:
i'm not fond of that. What if one comes with a contribution with some other markers because he/she uses a specific editor? It would become a bit messy.

eaaltonen replied:
>  one comes with a contribution with some other markers because he/she uses a specific editor?

This is a possibility. Of course all comment blocks should be valid Doxygen.

I think this comes down to a value between judgement between the rendered documentation this provides and the possibility you highlighted.

jenswi-linaro replied:
I'd prefer giving priority to the humans reading the source, that is, skip html markups.

etienne-lms replied:
`<translate lang=french>Je suis d'accord</translate>`

eaaltonen replied:
1 for, 2 against.

What of the
```
 * For reference, see
 * <a href="https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/tee/tee_core.c?h=v5.10#n26">
 * tee.c in Linux kernel</a>.
```
shall I remove it completely?

etienne-lms replied:
yes please.

jforissier commented:
`teecli` and the `cli/` source folder are too vague for my taste. This is pkcs#11 stuff, right? So how about using the `ck` prefix like for other related components? `teeckcli` and `ckcli/` perhaps?

eaaltonen commented:
> `teecli` and the `cli/` source folder are too vague for my taste. This is pkcs#11 stuff, right? So how about using the `ck` prefix like for other related components? `teeckcli` and `ckcli/` perhaps?

Ultimately it comes down to how you want to structure the project, and a bit of guessing about the future.

For example the _Network Security Services_ library build two somewhat related cli tools, `modutil` and `certutil`. If this project would be building mutliple utilities in a similar manner, they could both be build under `cli`, or in separate directories.

I would be fine with `teeckcli` and `ckcli/`.

eaaltonen commented:
> teeckcli and ckcli/ perhaps?

@etienne-lms: do you agree these should be the names? I'll continue next week.

etienne-lms commented:
Ok with **ckcli/** and **teeckcli**.

eaaltonen commented on libckteeaclc/src/group.c:
```diff
@@ -0,0 +1,50 @@
+
+enum rv_groupmember ckteeaclc_current_user_is_member_of(gid_t group)
+{
+	char username[L_cuserid];
```
Oops, slipped. I'm targeting two repositories; it's more work than usual. Will be included in the next push.

eaaltonen commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -12,12 +12,60 @@
+ * <a href="https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/tee/tee_core.c?h=v5.10#n26">
+ * tee.c in Linux kernel</a>.
+ */
+extern uuid_t KERNEL_NAMESPACE_UUID;
```
This is trying to give the impressions of a constant. (_although it's actually set in_ `ckteeaclc_lib_init`). It's supposed to be used like a read only constant.

etienne-lms replied:
Macros may not be constants, there are just... macros.
Here, really prefer a lower case label.

eaaltonen commented on libckteec/CMakeLists.txt:
```diff
@@ -53,6 +53,8 @@ target_compile_definitions (ckteec
 target_include_directories(ckteec
 	PUBLIC include
+	PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+	$<INSTALL_INTERFACE:include>
```
This is based on example in https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_INCLUDE_DIRECTORIES.html
quote
> Relative paths are allowed within the INSTALL_INTERFACE expression and are interpreted relative to the installation prefix.

This is also consistent with the example in the page about [creating relocatable packages](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#creating-relocatable-packages).

etienne-lms replied:
ok

eaaltonen commented on libckteeaclc/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+################################################################################
+target_include_directories(ckteeaclc PUBLIC
+	$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+	$<INSTALL_INTERFACE:include>
```
Wouldn't change based on the docs.

etienne-lms replied:
ok

eaaltonen commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -0,0 +1,54 @@
+
+/*
+ * Definitions for configuring and using Access Control List (ACL)
+ * authentication on OP-TEE tokens.
```
Went for "on PKCS#11 tokens provided by OP-TEE PKCS11 Trusted Application (TA)."

eaaltonen commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -0,0 +1,54 @@
+ * ckteeaclc_user_is_member_of() - Check if `user` is a member in `group`.
+ *
+ * @param user username string.
+ * @param group	to check membership of.
```
Is there any reference for the indent convention? I took a look at
https://www.kernel.org/doc/html/v5.11/doc-guide/kernel-doc.html but it's not that one.

eaaltonen replied:
@etienne-lms: any comment?

etienne-lms replied:
I have no strong opinion on the indentation here.
But at least split the argument label from the description, at least:
```patch
- * @param user username string.
+ * @param user Username string.
- * @param group	to check membership of.
+ * @param group Group ID to check membership of.
```

eaaltonen commented on libckteeaclc/src/group.c:
```diff
@@ -0,0 +1,50 @@
+		}
+	}
+
+	for (int i = 0; i < grouplistsize; ++i) {
```
done

eaaltonen commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -0,0 +1,54 @@
+ * ckteeaclc_current_user_is_member_of() - Check if the effective user id of
+ * the process is a member in `group`.
+ *
+ * @param group to check membership of.
```
4 spaces? Can I use that as a rule?

eaaltonen replied:
@etienne-lms : any comment?

eaaltonen commented on libckteec/src/invoke_ta.c:
```diff
@@ -259,27 +254,46 @@ CK_RV ckteec_invoke_init(void)
+			return CKR_ARGUMENTS_BAD;
 		}
 	}
+	return ckteec_invoke_init_with_login(login_method, login_gid);
```
added

eaaltonen commented on ckteecpp/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+target_include_directories(
+    ${PROJECT_NAME}
+    INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>
```
AFAICT, no

etienne-lms replied:
ok

eaaltonen commented on ckteecpp/include/ckteecpp/GroupPKCS11Login.hpp:
```diff
@@ -0,0 +1,76 @@
+	/**
+	 * Try to login as <tt>userType</tt> user.
+	 * @param pkcs11Session existing PKCS11Session
+	 * @param userType usually <tt>CK_USER</tt> or <tt>CKU_SO</tt>
```
For constructors, https://isocpp.org/wiki/faq/exceptions#ctors-can-throw recommends
> The best way to signal constructor failure is therefore to throw an exception.

More specific exceptions is something that I imagine would be added in the future.

Destructors can't throw so that's why I've added the construtor's that accept a `CK_RV &` so that the `CK_RV` doesn't just disappear from the face of the earth. It leaves the caller the possibility of reacting to the return value, if needed.

Added missing `@param`.

etienne-lms replied:
resolved

eaaltonen commented:
> teecli and the cli/ source folder are too vague for my taste. This is pkcs#11 stuff, right? So how about using the ck prefix like for other related components? teeckcli and ckcli/ perhaps?

I did the renaming, but I suppose you'll want to have those changes squashed to the respective commits. I decided to update the PR with what I have now.

etienne-lms commented:
@eaaltonen, I am ok if you squash the fixup commits into a nice commit series.

eaaltonen commented:
@etienne-lms && @jforissier : I've squashed the renaming into the existing commits. I also mentioned you on a couple of things on which I would like to have further input.

jbech-linaro commented on CMakeLists.txt:
```diff
@@ -37,3 +42,9 @@ add_subdirectory (libteec)
 add_subdirectory (public)
 add_subdirectory (libckteec)
+find_package (PkgConfig REQUIRED)
+pkg_check_modules(uuid REQUIRED IMPORTED_TARGET uuid)
```
Do you know what package to install on Ubuntu 18.04 to get this? The current build error seems to be due the "uuid" thing.

eaaltonen replied:
IIRC, the Ubuntu 18.04 version was affected by [this bug](https://github.com/karelzak/util-linux/issues/683).

That being said, `uuid-dev` should bring the required package.

jbech-linaro replied:
That is what I tried the other day before putting the comment here:
```
$ sudo apt install uuid-dev
[sudo] password for optee:
Reading package lists... Done
Building dependency tree
Reading state information... Done
uuid-dev is already the newest version (2.31.1-0.4ubuntu3.7).
0 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.

$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.5 LTS
Release:        18.04
Codename:       bionic
```
But as we can  see our testing still complains (see [this log](https://optee.mooo.com:5000/logs/OP-TEE/optee_client/259/577658896/e2cbaf6280eb033ce17bef0f42290eb5317a2670#end-01-hikey6220-build) for example).

jbech-linaro replied:
If we merge this patch before that has been sorted out, then all subsequent PR's will break due to this.

etienne-lms replied:
Is there an alternative to uuid-dev?

eaaltonen replied:
Looking at the log
```
16971:  sed -e 's#@@STAGING_SUBDIR@@#aarch64-buildroot-linux-gnu/sysroot#' -e 's#@@TARGET_CFLAGS@@#-D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Os#' -e 's#@@TARGET_CXXFLAGS@@#-D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Os#' -e 's#@@TARGET_FCFLAGS@@#-Os#' -e 's#@@TARGET_LDFLAGS@@##' -e 's#@@TARGET_CC@@#bin/aarch64-linux-gnu-gcc#' -e 's#@@TARGET_CXX@@#bin/aarch64-linux-gnu-g++#' -e 's#@@TARGET_FC@@#bin/aarch64-linux-gnu-gfortran#' -e 's#@@CMAKE_SYSTEM_PROCESSOR@@#aarch64#' -e 's#@@TOOLCHAIN_HAS_FORTRAN@@#0#' -e 's#@@CMAKE_BUILD_TYPE@@#Release#' /home/optee/devel/hikey/buildroot/support/misc/toolchainfile.cmake.in > /home/optee/devel/hikey/out-br/host/share/buildroot/toolchainfile.cmake
.
.
.
 27190:  (mkdir -p /home/optee/devel/hikey/out-br/build/optee_client_ext-1.0/ && cd /home/optee/devel/hikey/out-br/build/optee_client_ext-1.0/ && rm -f CMakeCache.txt && PATH="/home/optee/devel/hikey/out-br/host/bin:/home/optee/devel/hikey/out-br/host/sbin:/usr/bin/:/home/optee/devel/ibart/ibartenv/bin:/bin:/sbin:/snap/bin:/usr/bin:/usr/games:/usr/local/bin:/usr/local/games:/usr/local/sbin:/usr/sbin:/home/optee/bin"  /snap/bin/cmake /home/optee/devel/hikey/out-br/build/optee_client_ext-1.0/ -DCMAKE_TOOLCHAIN_FILE="/home/optee/devel/hikey/out-br/host/share/buildroot/toolchainfile.cmake" -DCMAKE_INSTALL_PREFIX="/usr" -DCMAKE_COLOR_MAKEFILE=OFF -DBUILD_DOC=OFF -DBUILD_DOCS=OFF -DBUILD_EXAMPLE=OFF -DBUILD_EXAMPLES=OFF -DBUILD_TEST=OFF -DBUILD_TESTS=OFF -DBUILD_TESTING=OFF -DBUILD_SHARED_LIBS=ON  -DCFG_TA_TEST_PATH=y )
```
and
https://git.busybox.net/buildroot/tree/support/misc/toolchainfile.cmake.in
line
`set(ENV{PKG_CONFIG_SYSROOT_DIR} "${RELOCATED_HOST_DIR}/@@STAGING_SUBDIR@@")`

I'm soft inferring that you would need to install a `aarch64` version of uuid-dev to the `STAGING_SUBDIR`, preferably at least version `2.34`, since the Makefile verifies that due to the mentioned UUID v5 bug.

I am not at all familiar with buildroot so I don't know what should be modified.

etienne-lms commented on libckteeaclc/src/cktee_uuid.c:
```diff
@@ -0,0 +1,34 @@
+int ckteeaclc_group_acl_uuid(char *uuid_buf, gid_t group)
+{
+	char gid_buf[CKTEEACLC_L_UUID];
+	int rv = sprintf(gid_buf, "gid=%x", group);
```
prefer `snprintf()`, the same way you use `strnlen()` below.

eaaltonen replied:
fixup commit

etienne-lms commented on ckteecpp/include/ckteecpp/BadLoginGroupArgument.hpp:
```diff
@@ -0,0 +1,33 @@
+ */
+
+#ifndef OPTEE_CLIENT_BADLOGINGROUPARGUMENT_HPP
+#define OPTEE_CLIENT_BADLOGINGROUPARGUMENT_HPP
```
s/OPTEE_CLIENT_/CKTEECPP_/g

eaaltonen replied:
fixup commit

etienne-lms reviewed (commented):
some minor comments

etienne-lms reviewed (commented):
Some minor coding style comments.
As for the implementation, it looks good to me, AFAICT from my cpp knowledge.

eaaltonen commented on ckcli/src/teeckcli.cpp:
```diff
@@ -0,0 +1,240 @@
+	}
+	ckteec_invoke_init_login_group(soGroup);
+
+	CK_UTF8CHAR_PTR pSoPin = (CK_UTF8CHAR_PTR) "";
```
Going by what I was told about proper C++ casting for this should be
```
CK_UTF8CHAR_PTR pSoPin = reinterpret_cast<CK_UTF8CHAR_PTR>(const_cast<char*>(""));
```

But the C style cast does the same and looks much easier to the eye (_for me_).

etienne-lms replied:
Ok. I don't have a strong opinion on this.
I fine with the current proposal.

etienne-lms replied:
resolved

eaaltonen commented on libckteeaclc/src/cktee_uuid.c:
```diff
@@ -17,7 +17,7 @@ int ckteeaclc_lib_init(void)
-	int rv = sprintf(gid_buf, "gid=%x", group);
+	int rv = snprintf(gid_buf, CKTEEACLC_L_UUID, "gid=%x", group);
 
 	if (rv < 0)
```
The `.h `stated `* @return 0 on success, otherwise a negative number is returned in case of failure.` here, so I changed to
```
	if (rv < 0)
		return rv;
	if (rv >= CKTEEACLC_L_UUID)
		return -1;
```
if that's allright.

etienne-lms replied:
fine with me. thanks.

eaaltonen commented on ckcli/src/teeckcli.cpp:
```diff
@@ -0,0 +1,240 @@
+		true, // show help if requested
+		"teeckcli "+ tee::cli::TEECKCLI_VERSION); // version string
+
+	#if DEBUG
```
If I just put a naive stub
```
void printArgs(const map<string, docopt::value> &args) {
#if DEBUG
	for (auto const& arg : args)
	{
		cout << arg.first << " " << arg.second << endl;
	}
#endif
}
```
I get `error: unused parameter 'args' [-Werror=unused-parameter]` when compiling.

etienne-lms replied:
maybe:
```c
void printArgs(const map<string, docopt::value> &args) {
#if DEBUG
	for (auto const& arg : args)
	{
		cout << arg.first << " " << arg.second << endl;
	}
#else
	(void)args;
#endif
}

eaaltonen replied:
Yeah that works.

eaaltonen commented on ckcli/src/teeckcli.cpp:
```diff
@@ -0,0 +1,240 @@
+			return initTokenWithPin(tokenLabel, slotId, soPin);
+		} else {
+			string soGroup = args["--set-so-group"].asString();
+			gid_t soGid;
```
Added ` = 0 ` for all `gid_t`s in `ckcli/src/teeckcli.cpp`

etienne-lms replied:
thanks

eaaltonen commented on libckteec/include/ckteec_extensions.h:
```diff
@@ -0,0 +1,71 @@
+} // extern "C"
+#endif
+
+#endif //OPTEE_CLIENT_CKTEEC_EXTENSIONS_H
```
fixup

eaaltonen commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -0,0 +1,114 @@
+} // extern "C"
+#endif
+
+#endif //OPTEE_CLIENT_CKTEEACLC_H
```
fixup

etienne-lms commented on ckteecpp/include/ckteecpp/GroupLoginCryptoki.hpp:
```diff
@@ -0,0 +1,78 @@
+	{
+		switch (ckteeaclc_current_user_is_member_of(loginGroup)) {
+			case NOT_MEMBER:
+				char username[L_cuserid];
```
```c
				char username[L_cuserid] = { 0 };
```

eaaltonen replied:
That would result in a `ckteecpp/include/ckteecpp/GroupLoginCryptoki.hpp:65:9: error: jump to case label`.

It will have to go to the top of the function the, although it's not used in the other cases. I pushed a fixup commit, please ack whether it's fine.

eaaltonen commented:
Squashed the queue of pending `fixup` commits with `git rebase -i --autosquash d1ae57555d102bc84b1c0761098ced6718a1f785`. No conflicts.

etienne-lms commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -13,12 +13,57 @@
+ */
+extern uuid_t kernel_namespace_uuid;
+/*
+ * making KERNEL_NAMESPACE_UUID const would an initializer list or string,
```
I think this needs some rephrasing.

etienne-lms commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -13,12 +13,57 @@
+/**
+ * Required length for UUID char buffers
+ */
+#define CKTEEACLC_L_UUID 48
```
Why 48 if max length is 43 bytes as stated right above?

eaaltonen replied:
I thought it might be good to round up. I chose the nearest multiple of 8 bytes. It does leave some room for additional login methods, so that's a potential benefit.

etienne-lms commented on libckteeaclc/include/ckteeaclc.h:
```diff
@@ -13,12 +13,57 @@
+ * @param group Group id to encode for login.
+ * @return 0 on success, otherwise a negative number is returned in case of failure.
+ */
+int ckteeaclc_group_acl_uuid(char *uuid_buf, gid_t group);
```
if `uuid_buf` is expected a `char` array of `CKTEEACLC_L_UUID`, this should be:
`int ckteeaclc_group_acl_uuid(char uuid_buf[CKTEEACLC_L_UUID], gid_t group);`
no?

eaaltonen replied:
Oh definitely :+1:. I wasn't aware that was an option. I've just seen so many complaints about C arrays losing their length that I didn't consider it.

etienne-lms commented on libckteec/include/ckteec_extensions.h:
```diff
@@ -0,0 +1,71 @@
+ */
+
+#ifndef OPTEE_CLIENT_CKTEEC_EXTENSIONS_H
+#define OPTEE_CLIENT_CKTEEC_EXTENSIONS_H
```
s/OPTEE_CLIENT_CKTEEC_EXTENSIONS_H/CKTEEC_EXTENSIONS_H/

etienne-lms commented on ckcli/src/TeeCkCliConstants.hpp:
```diff
@@ -0,0 +1,19 @@
+ */
+
+#ifndef OPTEE_CLIENT_TEECKCLICONSTANTS_HPP
+#define OPTEE_CLIENT_TEECKCLICONSTANTS_HPP
```
s/OPTEE_CLIENT_TEECKCLICONSTANTS_HPP/CKCLI_TEECKCLICONSTANTS_HPP/g

etienne-lms commented on ckcli/Makefile:
```diff
@@ -0,0 +1,85 @@
+
+DEBUG		?= 0
+ifeq ($(DEBUG),1)
+CXXFLAGS	+= -DDEBUG -O0 -g
```
Should set `-DDEBUG=1` to ensure `#if DEBUG` do match in source file.

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "Use separate generator expression for lib targets"
maybe change header line to: `cmake: use separate generator expression for lib targets`
* "Add libckteeaclc library"


Comments for commit "Add function to encode a group login based uuid". There is also this uuid tool dependency issue to be addressed.

Minor comment for commit "Add function to init TEE with login methods". I think this commit should be rebased. I would suggest you rebase this whole P-R on master branch top.

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "Add ckteecpp Interface library"
* "Write version to file to skip unnecessary builds"

eaaltonen commented:
Rebased on `master`. Three-way merge on "Add function to resolve name to gid". Looks ok and builds.

eaaltonen commented:
Added `Acked-by` and `Reviewed-by` lines to respective commits. Removed most HTML markup tags from `ckteecpp` files. Added fixup commits or commented on reasons for other commented issues

eaaltonen commented on ckteecpp/include/ckteecpp/cktee.hpp:
```diff
@@ -0,0 +1,39 @@
+ * <li> Cryptoki
+ * <li> PKCS11Session
+ * <li> PKCS11Login
+ * </ul>
```
I'm requesting to retain this block of markup. This part server only as documentation, and presenting lists is much clearer using the list format. IMO this is also not an eyesore even in plaintext.

eaaltonen commented:
Squashed last weeks queue of fixup commits, with `git rebase -i --autosquash 2c812c5b4f81a97687fb42196c1225545cbeb7e8`. Nothing to note.

etienne-lms commented:
@eaaltonen , this P-R is stalled because of the uuid lib dependency issue. Maybe it could be addressed using a br-ext/ package in OP-TEE build repo. Or maybe the easier way could be to dump the lib inside optee_client, or maybe dump only the lib header file if it is the part currently missing. Any idea?
